File: tests/omnia/data/cache/SimpleLongCacherTest.java
package omnia.data.cache;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.function.LongSupplier;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class SimpleLongCacherTest {

  @Test
  public void value_didReturnSuppliedValue() {
    long testValue = 132L;
    CachedLong testSubject = new SimpleLongCacher(() -> testValue);

    assertThat(testSubject.value()).isEqualTo(testValue);
  }

  @Test
  public void value_twice_didReturnSuppliedValue() {
    long testValue = 132L;
    CachedLong testSubject = new SimpleLongCacher(() -> testValue);

    testSubject.value();

    assertThat(testSubject.value()).isEqualTo(testValue);
  }

  @Test
  public void new_didNotInvokeSupplier() {
    LongSupplier supplier = setUpMockSupplier();
    new SimpleLongCacher(supplier);

    verify(supplier, never()).getAsLong();
  }

  @Test
  public void invalidate_didNotSupplier() {
    LongSupplier supplier = setUpMockSupplier();
    CachedLong testSubject = new SimpleLongCacher(supplier);

    testSubject.invalidate();

    verify(supplier, never()).getAsLong();
  }

  @Test
  public void value_didInvokeSupplier() {
    LongSupplier supplier = setUpMockSupplier();
    CachedLong testSubject = new SimpleLongCacher(supplier);

    testSubject.value();

    verify(supplier).getAsLong();
  }

  @Test
  public void value_didInvokeSupplierOnce() {
    LongSupplier supplier = setUpMockSupplier();
    CachedLong testSubject = new SimpleLongCacher(supplier);

    testSubject.value();
    testSubject.value();

    verify(supplier, times(1)).getAsLong();
  }

  @Test
  public void value_thenInvalidate_thenValue_didInvokeSupplierTwice() {
    LongSupplier supplier = setUpMockSupplier();
    CachedLong testSubject = new SimpleLongCacher(supplier);

    testSubject.value();
    testSubject.invalidate();
    testSubject.value();

    verify(supplier, times(2)).getAsLong();
  }

  @Test
  public void new_withNullSupplier_didThrowException() {
    assertThrows(NullPointerException.class, () -> new SimpleLongCacher(null));
  }

  private static LongSupplier setUpMockSupplier() {
    LongSupplier supplier = mock(LongSupplier.class);
    when(supplier.getAsLong()).thenReturn(132L);
    return supplier;
  }
}

File: tests/omnia/data/cache/SimpleLongCacherTest.kt
package omnia.data.cache

import com.google.common.truth.Truth.assertThat
import kotlin.test.Test

import org.mockito.Mockito
import org.mockito.kotlin.mock

class SimpleLongCacherTest {

  @Test
  fun value_didReturnSuppliedValue() {
    val testValue = 132L
    val testSubject: CachedLong = SimpleLongCacher { testValue }
    assertThat(testSubject.value()).isEqualTo(testValue)
  }

  @Test
  fun value_twice_didReturnSuppliedValue() {
    val testValue = 132L
    val testSubject: CachedLong = SimpleLongCacher { testValue }
    testSubject.value()
    assertThat(testSubject.value()).isEqualTo(testValue)
  }

  @Test
  fun new_didNotInvokeSupplier() {
    val supplier = setUpMockSupplier()
    SimpleLongCacher(supplier)
    Mockito.verify(supplier, Mockito.never()).invoke()
  }

  @Test
  fun invalidate_didNotSupplier() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedLong = SimpleLongCacher(supplier)
    testSubject.invalidate()
    Mockito.verify(supplier, Mockito.never()).invoke()
  }

  @Test
  fun value_didInvokeSupplier() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedLong = SimpleLongCacher(supplier)
    testSubject.value()
    Mockito.verify(supplier).invoke()
  }

  @Test
  fun value_didInvokeSupplierOnce() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedLong = SimpleLongCacher(supplier)
    testSubject.value()
    testSubject.value()
    Mockito.verify(supplier, Mockito.times(1)).invoke()
  }

  @Test
  fun value_thenInvalidate_thenValue_didInvokeSupplierTwice() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedLong = SimpleLongCacher(supplier)
    testSubject.value()
    testSubject.invalidate()
    testSubject.value()
    Mockito.verify(supplier, Mockito.times(2)).invoke()
  }

  companion object {

    private fun setUpMockSupplier(): () -> Long {
      val supplier = mock<() -> Long>()
      Mockito.`when`(supplier()).thenReturn(132L)
      return supplier
    }
  }
}